main.py

```diff
@@ -168,102 +181,115 @@ def sample_path(name:str, langs: Iterable[str]):
     return os.path.join(DATA_PATH, f'.sample.{name}.{languages}.gz')
 
 
-def compute_sample(name:str, columns:list[tuple[str,os.DirEntry]]):
+async def compute_sample(name:str, columns:list[tuple[str,os.DirEntry]]):
     langs = [lang for lang, _ in columns]
-    with ExitStack() as ctx, gzip.open(sample_path(name, langs), 'wb') as fout:
-        files = [ctx.enter_context(gzip.open(file.path, 'rb')) for _, file in columns]
+    with TemporaryFile() as tempfile, gzip.open(tempfile, 'wb') as fout:  # type: ignore[name-defined]
+        proc = await asyncio.subprocess.create_subprocess_exec([SAMPLE_PY, '-n', SAMPLE_SIZE] + [file.path for _, file in columns],
+            stdout=fout,
+            stderr=asyncio.subprocess.PIPE)
 
-        pairs = zip(*files)
-        
-        head, middle, tail = sample(10, pairs)
+        _, stderr = await proc.communicate()
+
+        if proc.returncode != 0:
+            raise Exception(f'sample.py returned {proc.returncode}: {stderr.decode()}')
+
+        tempfile.seek(0)
 
-        for pair in chain(head, middle, tail):
-            fout.write(b'\t'.join(line.rstrip(b'\n') for line in pair) + b'\n')
+        with open(sample_path(name, langs), 'wb') as fdest:
+            copyfileobj(tempfile, fdest)
 
 
 class FilterOutput(BaseModel):
     stdout: list[dict[str,str]]
     stderr: Optional[str]
 
+    def __init__(self, langs:list[str], stdout:bytes, stderr:Optional[bytes] = None):
+        super().__init__(
+            stdout=[dict(zip(langs, line.split('\t'))) for line in stdout.decode().split('\n')],
+            stderr=stderr.decode() if stderr is not None else None)
 
-def get_sample(name:str, filters:list[FilterStep]) -> FilterOutput:
+
+async def get_sample(name:str, filters:list[FilterStep]) -> AsyncIterator[FilterOutput]:
```

Comment:
I assume the reason for moving to `async` is that you are going to call a bunch of `get_sample()` in parallel and await for them?

Reply:
No, in this case it is so it can send the intermediate output from the filter steps already to the browser while it is still running the next steps. And so it can handle other requests in between because the default server is not threaded I think.